### PR TITLE
Feature/265 replace main menu icons 3

### PR DESCRIPTION
**Pull request by ckleis**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no
| Fixed issues  | chameleon-system/chameleon-system#265
| License       | MIT

This feature-Branch changes the main menu icons to font-awesome icons.

### CODE REVIEW

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+<div class="changelog">
+    - https://github.com/chameleon-system/chameleon-system/issues/265
+    - backend - main navigation: add new field icon_font_awesome
+</div>
+<?php
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+        'cms_tbl_conf_id' => TCMSLogChange::GetTableId('cms_tbl_conf'),
+        'name' => 'new_field',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Please cleanup this update by merging insert and update statement (there should be a single insert statement that creates the record, plus translations).

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        'length_set' => '',
+        'fieldtype_config' => '',
+        'restrict_to_groups' => '0',
+        'field_width' => '',
+        'position' => '0',
+        '049_helptext' => '',
+        'row_hexcolor' => '',
+        'is_translatable' => '0',
+        'validation_regex' => '',
+        'id' => 'd1956950-5815-f44e-9881-023e28e354c3',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Please use `TCMSLogChange::createUnusedRecordId()` to generate IDs.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    ])
+;
+TCMSLogChange::insert(__LINE__, $data);
+
+$query = "ALTER TABLE `cms_tbl_conf`
+                        ADD `icon_font_awesome` VARCHAR(255) NOT NULL COMMENT 'Icon Font Awesome css class: '";
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: "css" --> "CSS" (multiple times in this file)

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    ])
+    ->setWhereEquals([
+        'id' => 'd1956950-5815-f44e-9881-023e28e354c3',
+    ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+//      'name' => 'icon_font_awesome',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Please remove code that is commented out.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        'translation' => 'Icon Font Awesome css class',
+    ])
+    ->setWhereEquals([
+        'id' => 'd1956950-5815-f44e-9881-023e28e354c3',
+    ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: See above (cleanup update).

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        'length_set' => '',
+        'fieldtype_config' => '',
+        'restrict_to_groups' => '0',
+        'field_width' => '',
+        'position' => '0',
+        '049_helptext' => '',
+        'row_hexcolor' => '',
+        'is_translatable' => '0',
+        'validation_regex' => '',
+        'id' => '63cde82b-57a4-61d0-1444-e05638bf3aef',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: See above (generate ID)

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    ])
+    ->setWhereEquals([
+        'id' => '63cde82b-57a4-61d0-1444-e05638bf3aef',
+    ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+//      'name' => 'icon_font_awesome',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Please remove code that is commented out.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        ])
+        ->setWhereEquals([
+            'name' => $tableName
+        ])
+    ;
+    TCMSLogChange::update(__LINE__, $data);
+}
+
+
+//set new incons for modules
```
[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php]: changed

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        ])
+        ->setWhereEquals([
+            'id' => $row['id']
+        ])
+    ;
+    TCMSLogChange::update(__LINE__, $data);
+    $modulesWithEmptyIcon[] = $row['name'];
+}
+
+if (sizeOf($modulesWithEmptyIcon) > 0) {
```
[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php]: changed

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        ])
+        ->setWhereEquals([
+            'id' => $row['id']
+        ])
+    ;
+    TCMSLogChange::update(__LINE__, $data);
+    $tablesWithEmptyIcon[] = $row['name'];
+}
+
+if (sizeOf($tablesWithEmptyIcon) > 0) {
```
[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php]: changed

File: src/CoreBundle/private/library/classes/MenuItems/TCMSMenuItem_Module.class.php [hunk 11]
```diff
@@ -31,12 +31,12 @@ public function GetLink()
             $url = "javascript:CreateModalIFrameDialogCloseButton('".$url."',".TGlobal::OutHTML($this->data['width']).','.TGlobal::OutHTML($this->data['height']).');';
         }
 
-        $iconStyle = '';
-        $rightIconStyle = 'padding-left: 20px;';
-        if (array_key_exists('icon_list', $this->data) && !empty($this->data['icon_list'])) {
-            $iconStyle = ' style="background-image: url('.URL_CMS.'/images/icons/'.$this->data['icon_list'].'); background-repeat: no-repeat; background-position: left center; line-height: 17px;"';
+        if (array_key_exists('icon_font_awesome', $this->data) && !empty($this->data['icon_font_awesome'])) {
```
[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/private/library/classes/MenuItems/TCMSMenuItem_Module.class.php]: changed

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 3 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: replaced

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 4 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 5 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 7 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 6 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: How can I generate an ID?

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+
+if (false === $statement->execute()) {
+    return;
+}
+
+$tablesWithEmptyIcon = [];
+while (false !== $row = $statement->fetch()) {
+    $data = TCMSLogChange::createMigrationQueryData('cms_tbl_conf', 'en')
+        ->setFields([
+            'icon_font_awesome' => 'fas fa-sign-out-alt'  //standard icon
```
[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php]: Fallback standard icon in this update file removed

[the remark this answers is not in the document]
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452870.inc.php]: Text added to Font Awesome and its use in file UPGRADE-6.3.md

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+<div class="changelog">
+    - https://github.com/chameleon-system/chameleon-system/issues/265
+    - backend - main navigation: add new field icon_font_awesome
+</div>
+<?php
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+        'cms_tbl_conf_id' => TCMSLogChange::GetTableId('cms_tbl_conf'),
+        'name' => 'icon_font_awesome',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: On second thought I'd say we should use a more neutral term for this new field. After all we can put any CSS class declarations here and don't need to restrict to Font Awesome (and we might later switch to another font), so how about simply `css_class` with the help text explaining that this field is used to display the table as menu item?

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 6 above)
Comment (umulmrum) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: --> `TCMSLogChange::createUnusedRecordId()`

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 2 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Done.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 6 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: Done.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php (hunk 13 above)
Comment (ckleis) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1547452866.inc.php]: You're right, we can use a different font in the future or add another one, but it's not a general CSS class for the navigation link, but for the icon element <i> inside the link. So I would suggest to call the field "Icon Font CSS class", respectively "icon_font_css_class".

**Comment by umulmrum** [comment]
Thanks @ckleis

**Pull request merged**